homeassistant/components/scheduler/__init__.py

```diff
@@ -0,0 +1,432 @@
+"""Support for scheduling scenes."""
+from datetime import datetime, timedelta
+import logging
+from typing import TYPE_CHECKING, Callable, Dict, List, Optional
+from uuid import uuid4
+
+from dateutil.rrule import rrulestr
+import voluptuous as vol
+
+from homeassistant.components import websocket_api
+from homeassistant.const import CONF_ENTITY_ID, CONF_ID, MATCH_ALL
+from homeassistant.core import Context, callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.event import (
+    async_track_point_in_time,
+    async_track_state_change,
+)
+import homeassistant.util.dt as dt_util
+from homeassistant.util.json import load_json, save_json
+
+if TYPE_CHECKING:
+    from homeassistant.core import HomeAssistant, ServiceCall, State  # noqa
+
+_LOGGER = logging.getLogger(__name__)
+DOMAIN = "scheduler"
+
+PERSISTENCE = ".scheduler.json"
+EVENT = "scheduler_updated"
+
+CONF_ACTIVATION_CONTEXT = "activation_context"
+CONF_ACTIVATION_CONTEXT_ID = "activation_context_id"
+CONF_DURATION = "duration"
+CONF_END_DATETIME = "end_datetime"
+CONF_RRULE = "rrule"
+CONF_SCHEDULE_ID = "schedule_id"
+CONF_START_DATETIME = "start_datetime"
+
+SCHEDULE_UPDATE_SCHEMA = vol.All(
+    vol.Schema({CONF_ENTITY_ID: cv.entity_id, CONF_DURATION: int, CONF_RRULE: str}),
+    cv.has_at_least_one_key(CONF_ENTITY_ID, CONF_DURATION, CONF_RRULE),
+)
+
+WS_TYPE_SCHEDULER_CREATE_SCHEDULE = "scheduler/schedules/create"
+WS_TYPE_SCHEDULER_DELETE_SCHEDULE = "scheduler/schedules/delete"
+WS_TYPE_SCHEDULER_SCHEDULES = "scheduler/schedules"
+WS_TYPE_SCHEDULER_UPDATE_SCHEDULE = "scheduler/schedules/update"
+
+SCHEMA_WEBSOCKET_CREATE_SCHEDULE = websocket_api.BASE_COMMAND_MESSAGE_SCHEMA.extend(
+    {
+        vol.Required("type"): WS_TYPE_SCHEDULER_CREATE_SCHEDULE,
+        vol.Required(CONF_ENTITY_ID): cv.entity_id,
+        vol.Required(CONF_START_DATETIME): cv.datetime,
+        vol.Optional(CONF_RRULE): str,
+        vol.Optional(CONF_DURATION): int,
+    }
+)
+
+SCHEMA_WEBSOCKET_DELETE_SCHEDULE = websocket_api.BASE_COMMAND_MESSAGE_SCHEMA.extend(
```

Comment:
Please use the modern websocket API decorator 

https://github.com/home-assistant/home-assistant/blob/957e5018f4445cd83de01e3da07717f7afa23909/homeassistant/components/conversation/__init__.py#L107-L109